Minor fixes

hroest opened the pull request:
a series of minor fixes that addresses some memory leaks and a segfault in TOPPView

timosachsenberg commented on src/openms_gui/source/VISUAL/TOPPViewSpectraViewBehavior.cpp:
```diff
@@ -324,6 +324,14 @@ namespace OpenMS
           peak1d.setIntensity(cpeak.getIntensity());
           spectrum.push_back(peak1d);
         }
+        if (current_chrom.size() == 0) 
```
.empty() ?

hroest replied:
sounds good, I will change it

hroest commented:
@cbielow do you have a better idea for the TOPPView fix? Currently the issue is that the layer is empty and then adding the layer fails, which leads to a segfault when next trying to access the layer. My solution is to simply add at least one datapoint in case of empty chromatograms so that this never happens, instead an empty 1D view is shown (not such a bad outcome).

hroest commented on src/openms/source/FORMAT/XMLFile.cpp:
```diff
@@ -152,21 +154,38 @@ namespace OpenMS
+        // nothing to do here except free memory, as this exception is used to
+        // softly abort the parsing for whatever reason.
+        delete parser;
+        delete source;
```
@cbielow @mwalzer do you agree with this, I think this is a memory leak if we use the `EndParsingSoftly` to abort parsing but continue with the execution: in this case we should clean up memory as we have a mem leak otherwise

cbielow replied:
agreed, good catch!

A clean alternative to all these delete calls, would be RAII, i.e. either a custom class (as done in  XMLCleaner_) or an auto_ptr/std::unique_ptr. The former is not encouraged for C11 and I'm not sure about the broad support of std::unique_ptr for the platforms we run on. VS2010 does have it though.

hroest replied:
`std::auto_ptr` should be C++03 if my internet is correct ...

hroest replied:
or we can just use `boost::shared_ptr` like in most other places

hroest commented on src/openms/source/FORMAT/XMLFile.cpp:
```diff
@@ -152,21 +154,38 @@ namespace OpenMS
+        // clean up memory first, then re-throw
+        delete parser;
+        delete source;
+        throw;
```
this very likely aborts the program anyways but we should still be nice and clean up memory...

cbielow commented on src/openms_gui/source/VISUAL/TOPPViewSpectraViewBehavior.cpp:
```diff
@@ -324,6 +324,16 @@ namespace OpenMS
           spectrum.push_back(peak1d);
         }
+        
+        // Add at least one data point to the chromatogram, otherwise "addLayer" will fail and a segfault occurs later
```
could this also be handled in addLayer or during display? This looks like a workaround which does not solve the underlying problem, i.e. no error checking in later stages (which should be robust against these kind of things if used independenly)

hroest replied:
yes, but I am not experienced enough with the code to do that at the moment. It seems that 
` bool Spectrum1DCanvas::finishAdding_()` doesnt really deal with it at all but just outputs an error and returns false and then `TOPPViewSpectraViewBehavior::activate1DSpectrum` returns. This, however, does not seem to leave things in a good state.

https://github.com/OpenMS/OpenMS/blob/develop/src/openms_gui/source/VISUAL/Spectrum1DCanvas.cpp#L1025

hroest replied:
the next thing that happens is
```
Caught exception: 'Precondition failed' with message 'index < layers_.size() SpectrumCanvas::getLayer_(index) index overflow'
```

cbielow commented:
I also checked if deleting all these resources is the responsibility of the caller and not of Xerces, but this seems to be the case.

hroest commented:
> I also checked if deleting all these resources is the responsibility of the caller and not of Xerces, but this seems to be the case.

yes, this seems to be the case, so we need to make sure they are deleted in every code path

hroest commented:
@cbielow regarding the TOPPView part: 

the main issue is that we store multiple types of meta-information in the current layer, mainly whether it is a chromatogram layer and if so, where the actual data for the chromatograms are located. If we remove the layer, then we loose all of this info as well and we are in quite some trouble. Simply giving up and returning here https://github.com/OpenMS/OpenMS/blob/develop/src/openms_gui/source/VISUAL/TOPPViewSpectraViewBehavior.cpp#L225 is also not a good option:

```
if (widget_1d->canvas()->getLayerCount() == 0) return;
```

from a user perspective as then the program just draws an empty canvas and one has to close the current tab, re-enter the 1D view and start from scratch. To me, the current solution with a "dummy" layer seems best as this does not break any of the assumptions underlying the code and gives the user the expected experience. I agree its not pretty, but it seems the other options are worse.

hroest commented:
I have now added some new code, I did implement the check for the layer count while also adding the dummy layer. If the `addLayer` call fails for any other reason, we would now still avoid a segfault and just present the user with an empty canvas (still better than segfaulting).

I also fixed a segfault when reading empty chromatograms and added a test

timosachsenberg commented:
can this be merged?

cbielow commented on src/openms/include/OpenMS/FORMAT/HANDLERS/MzMLHandler.h:
```diff
@@ -693,24 +693,39 @@ namespace OpenMS
+        // repair size of array, accessing memory that is beyond int_size will lead to segfaults later
+        if (repair_array_length)
+        {
+          default_arr_length = std::min(int_size, rt_size);
```
this does not make sense.
We asserted already that int_size == rt_size.
Maybe
`default_arr_length  = int_size; // set to length of actual data (int_size and rt_size are equal, s.a.)`

cbielow commented on src/openms/source/FORMAT/XMLFile.cpp:
```diff
@@ -152,21 +154,38 @@ namespace OpenMS
       catch (const xercesc::XMLException & toCatch)
       {
-        throw Exception::ParseError(__FILE__, __LINE__, __PRETTY_FUNCTION__, "", String("XMLException: ") + StringManager().convert(toCatch.getMessage()));
+        delete parser;
```
opting for boost::shared_ptr here since ideally you would have
```delete parser;
parser = 0;
...```

do avoid double deletion etc... we can save a lot of code just using shared_ptr.

cbielow replied:
Or std::auto_ptr.. even though it is dicsouraged for C11+, .. so not ideal...

hroest commented:
@cbielow @timosachsenberg I think I adressed all comments and now use boost::shared_ptr which is easiest and supported throughout our codebase

timosachsenberg commented:
+1

timosachsenberg commented:
there are some failing tests

timosachsenberg commented:
Error: File not found (the file 'FileConverter_23.tmp' could not be found)
Cannot read input file given from parameter '-in1'!

hroest commented:
@timosachsenberg strange, on my test machine it works in debug mode. Is there a difference how warnings/errors are handled in debug or test mode?

```
$ ctest -R FileConverter_23 
Test project /home/hr/openmsall/builds/debug_build
    Start 1123: TOPP_FileConverter_23
1/2 Test #1123: TOPP_FileConverter_23 ............   Passed    0.55 sec
    Start 1124: TOPP_FileConverter_23_out
2/2 Test #1124: TOPP_FileConverter_23_out ........   Passed    0.06 sec

100% tests passed, 0 tests failed out of 2
```

in the worst case, we can also remove the test again but I find it a bit strange that it doesnt work ...

hroest commented:
or probably just a missing `PROPERTIES DEPENDS` :-) see e7fa426cd656fd259dcb3b27ff6e2aca89ccf1d1
